PhoneBook.Web/Classes/EditFormPage.cs

```diff
@@ -85,47 +74,47 @@ protected bool SaveObject()
       EnsurePostLoadInvoked();
       EnsureValidatableControlsInitialized();
 
-      bool isValid = DataSource.Validate();
-      Control firstInvalidControl = null;
+      var isValid = DataSource.Validate();
 
-      foreach (DataEditUserControl control in DataEditUserControls)
+      var firstInvalidControl = DataEditUserControls.FirstOrDefault(control => !control.Validate());
+      if (firstInvalidControl != null)
       {
-        if (! control.Validate())
-        {
-          isValid = false;
-          firstInvalidControl = control;
-          break;
-        }
+        isValid = false;
       }
-
+      
       if (isValid)
       {
-        foreach (DataEditUserControl control in DataEditUserControls)
-          control.SaveValues (false);
-
-        DataSource.SaveValues (false);
-        _isSaved = true;
-        // transaction will be committed by caller or using auto commit.
+        SaveValidObject();
       }
       else
       {
-        TabbedMultiView multiView = UserControlMultiView;
-        if (multiView != null)
+        var multiView = UserControlMultiView;
+        if (multiView == null)
+          return false;
+
+        for (Control control = firstInvalidControl; control != null; control = control.Parent)
         {
-          for (Control control = firstInvalidControl; control != null; control = control.Parent)
-          {
-            if (control is TabView)
-            {
-              multiView.SetActiveView ((TabView) control);
-              break;
-            }
-          }
+          if (!(control is TabView))
+            continue;
+
+          multiView.SetActiveView ((TabView) control);
+          break;
         }
       }
 
       return isValid;
     }
 
+    private void SaveValidObject ()
```

Comment:
Since Save is being factored out, both save-operations should be factored out (here and in Unload). Instead, the flag and the "valid" bit (+ the comment) would remain at the call site.